### PR TITLE
Policies

### PR DESCRIPTION
**Pull request by frankmer**
- Implemented the 1st part of the Policy Page and all required services.
 - Added tests for the Policies.
 - Implemented all required mocks and updated some existing mocks needed for the tests.
 - Updated other failed tests due to the mock updates.

### CODE REVIEW

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      }
+    </div>
+    <div class="action-detail-card">
+      <span
+        class="detail-row"
+        [innerHTML]="policyService.selectedActionDetail()?.desc">
+      </span>
+    </div>
+    @if (!isEditMode()) {
+    } @else if (policyService.selectedActionDetail()?.value) {
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html]: This logic is confusing. You can write <br>
```
@if (isEditMode()){
  @if (policyService.selectedActionDetail()?.value) {
  ...
} @ else {
  @switch
...
}
```

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          <mat-label i18n>Value</mat-label>
+          <mat-select
+            [value]="policyService.selectedAction()?.value"
+            (selectionChange)="policyService.updateSelectedActionValue($event.value)">
+            @for (option of policyService.selectedActionDetail()?.value; track option) {
+              <mat-option [value]="option">{{ option }}</mat-option>
+            }
+          </mat-select>
+        </mat-form-field>
+      }
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html]: I understand the idea and the app-selector-buttons are even nicer than a selection. However, it might be confusing that the layout differs only based on the selection options. For consistency, I would prefer a unique presentation regardless of the number of selection options. </br>

Additionally, **multi select** is not covered here (e.g. challenge_response)

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    } @else {
+      @switch (policyService.selectedActionDetail()?.type) {
+        @case ("bool") {
+          <!-- <app-bool-select-buttons
+            (onSelectString)="policyService.updateSelectedActionValue($event)"
+            [initialValue]="policyService.selectedAction()?.value ?? 'true'" /> -->
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html]: remove unused code

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            (ngModelChange)="policyService.updateSelectedActionValue($event)"
+            class="action-value-textarea"></textarea>
+        }
+        @default {
+          <p i18n>Unknown action type: {{ policyService.selectedActionDetail()?.type }}</p>
+        }
+      }
+    }
+    @if (!isEditMode()) {
+    } @else if (selectedActionIsAlreadyAdded()) {
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-detail/action-detail.component.html]: Why inverting the logic here?

```
@if (isEditMode()) {
  @if (selectedselectedActionIsAlreadyAdded()) {
  ...
  } @else {...}
}
```

File: privacyidea/static_new/src/app/services/policies/policies.service.ts [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+    url: `${this.policyBaseUrl}defs`,
+    method: "GET",
+    headers: this.authService.getHeaders()
+  }));
+
+  readonly allPoliciesRecource = httpResource<PiResponse<PolicyDetail[]>>(() => ({
+    url: `${this.policyBaseUrl}`,
+    method: "GET",
+    headers: this.authService.getHeaders()
+  }));
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/services/policies/policies.service.ts]: Can you please add a condition here to only load the resources if we are on the policy route? And additionally a check if the user has the rights to get the policies would be nice.

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-selector/action-selector.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  //   const input: HTMLInputElement = fixture.nativeElement.querySelector(".search-input");
+
+  //   input.value = "test";
+  //   input.dispatchEvent(new Event("input"));
+  //   TestBed.flushEffects();
+  //   fixture.detectChanges();
+
+  //   const actionFilter = policyServiceMock.actionFilter();
+  //   expect(actionFilter).toBe("test");
+  // });
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-selector/action-selector.component.spec.ts]: remove unused code

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-selector/action-selector.component.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  imports: [CommonModule, FormsModule, MatIconModule, SelectorButtons, MatTooltipModule],
+  templateUrl: "./action-selector.component.html",
+  styleUrls: ["./action-selector.component.scss"]
+})
+export class ActionSelectorComponent {
+  policyService: PolicyServiceInterface = inject(PolicyService);
+  testngModelChange(event: string) {
+    this.policyService.actionFilter.set(event);
+    throw Error(event);
+  }
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/action-selector/action-selector.component.ts]: I'm not sure what this code does, but we should avoid any testing code in the real component

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      </div>
+    } @else {
+      <textarea
+        class="description-textarea"
+        [ngModel]="policyService.selectedPolicy()?.description"
+        (ngModelChange)="updatePolicyDescription($event)"
+        placeholder="(optional) Enter a description for this policy..."
+        cdkTextareaAutosize
+        cdkAutosizeMinRows="3"
+        cdkAutosizeMaxRows="6"></textarea>
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html:34]: Can you use the same font as in the remaining UI here?

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-priority/policy-priority.component.html [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  @if (!editMode()) {
+    <div class="description-text">
+      {{ policyService.selectedPolicy()?.priority }}
+    </div>
+  } @else {
+    <input
+      type="number"
+      class="description-textarea"
+      [ngModel]="policyService.selectedPolicy()?.priority"
+      (ngModelChange)="updatePolicyPriority($event)" />
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-priority/policy-priority.component.html:30]: Can you align the input on the same height as the Priority text? Seems like the input has a top margin.

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-priority/policy-priority.component.scss [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  background-color: transparent;
+  border: none;
+  border-bottom: 1px solid var(--mat-sys-outline-variant);
+  border-radius: 0;
+  box-sizing: border-box;
+  font-size: 1rem;
+  padding: var(--half-global-padding) 0;
+  resize: vertical;
+  width: 100%;
+}
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-priority/policy-priority.component.scss:23]: If these styles are used multiple times in the policy components you could also define them only once int the policies.component.scss. Otherwise another name fitting this component would be nice.

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/selector-buttons/selector-buttons.component.scss [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+// .bool-select-button {
+//   background-color: var(--mat-sys-surface-container);
+//   border: none;
+//   border-radius: 0;
+//   cursor: pointer;
+//   flex: 1;
+//   margin: 0;
+//   padding: var(--half-global-padding) var(--three-quarters-global-padding);
+//   text-align: center;
+// }
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/selector-buttons/selector-buttons.component.scss]: remove unused code

File: privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-additional/conditions-additional.component.html [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+              } @else {
+                <mat-icon>close</mat-icon>
+              }
+            }
+          </div>
+          <div>{{ condition[0] }}</div>
+          <div>{{ condition[1] }}</div>
+          <div>{{ condition[2] }}</div>
+          <div>{{ condition[3] }}</div>
+          <div>{{ condition[5] }}</div>
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-additional/conditions-additional.component.html:72]: for readability a dict instead of a list for the condition type would be nice. Hence, we can write something like `condition.section`.

File: privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-additional/conditions-additional.component.html [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            </button>
+          </div>
+        </div>
+      } @else {
+        <div class="add-condition-button-container">
+          <button
+            mat-stroked-button
+            (click)="showAddConditionForm.set(true)">
+            <mat-icon>add</mat-icon>
+            Create new condition
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-additional/conditions-additional.component.html]: Create a New Condition?

File: privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-nodes/conditions-nodes.component.html [hunk 14]
```diff
@@ -0,0 +1,136 @@
+@if (isEditMode() || policyService.selectedPolicyHasNodeConditions()) {
+  <div class="nodes-container">
+    <h2
+      class="nodes-title"
+      i18n>
+      Nodes
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-nodes/conditions-nodes.component.html]: I think this title not really fits to the content. These conditions are not really related and only the first one is related to the nodes. To be honest I also have no other good suggestion, maybe something like system / environment / generic / miscellaneous.

File: privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-nodes/conditions-nodes.component.ts [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+  selectedPolicyName = computed(() => this.selectedPolicy?.name || "");
+  availablePinodesList = computed(() => this.systemService.nodes().map((node) => node.name));
+  selectedPinodes = computed<string[]>(() => this.selectedPolicy()?.pinode || []);
+  selectedUserAgents = computed(() => this.policyService.selectedPolicy()?.user_agents || []);
+  selectedValidTime = computed(() => this.policyService.selectedPolicy()?.time || "");
+  selectedClient = computed(() => this.policyService.selectedPolicy()?.client || "");
+  isAllNodesSelected = computed(() => this.selectedPinodes().length === this.availablePinodesList().length);
+
+  // Placeholder for available user agents
+  availableUserAgents = signal<string[]>(["Mozilla Firefox", "Google Chrome", "Microsoft Edge"]);
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-nodes/conditions-nodes.component.ts]: These should be the user agents of our plugins. (see old web ui for a complete list)

File: privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-user/conditions-user.component.scss [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+  flex-shrink: 0;
+}
+
+.full-width {
+  width: 100%;
+}
+
+.add-user-input {
+  width: 100%;
+  // ecken rechts sollen eckig sein
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/conditions-tab/conditions-user/conditions-user.component.scss]: comments should be in english

File: privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.html [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+                {{ newPolicyScope().toLocaleUpperCase() }}
+              </span>
+            }
+          } @else if (policy()) {
+            @if (isEditingPolicy(policy()!.name) && policyService.isScopeChangeable(selectedPolicy()!)) {
+              <app-horizontal-wheel
+                [initialValue]="selectedPolicy()!.scope"
+                [values]="policyService.allPolicyScopes"
+                (onSelect)="selectPolicyScope($event)"
+                (click)="$event.stopPropagation()" />
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.html]: hm the appearance of this wheel might be nice. Regarding usability, I would personally prefer a simple selection that allows me to see all available scopes at once.

File: privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.ts [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+    }
+  }
+
+  handleCollapse(panel: MatExpansionPanel, policyName: string | undefined) {
+    if (!this.policyIsSelected(policyName)) {
+      return;
+    }
+    if (this.isNew()) {
+      if (this.policyService.isPolicyEdited()) {
+        if (confirm("Are you sure you want to discard the new policy? All changes will be lost.")) {
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.ts:119]: In other components we use a dialog for these confirmations. For consistency, I would use a dialog here as well. 

Same for the following confirm() s

File: privacyidea/static_new/src/app/components/token/container-details/container-details-actions/container-details-actions.component.spec.ts [hunk 19]
```diff
@@ -75,14 +76,18 @@ describe("ContainerDetailsActionsComponent", () => {
...
     expect(compiled.textContent).toContain("Container Actions");
   });
 
   it("should not render actions if no action is allowed", () => {
-    mockAuthService.jwtData.set({...component.authService.jwtData(), rights: []});
+    authServiceMock.actionAllowed.mockImplementation((asd) => {
+      console.warn(asd);
+      return false;
+    });
+    // TestBed.flushEffects();
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/token/container-details/container-details-actions/container-details-actions.component.spec.ts]: What was wrong with the original implementation?

File: privacyidea/static_new/src/app/components/token/container-details/container-details-token-table/container-details-token-table.component.ts [hunk 20]
```diff
@@ -224,6 +219,51 @@ export class ContainerDetailsTokenTableComponent {
...
+          if (result) {
+            this.containerService.removeAll(this.containerSerial()).subscribe({
+              next: () => {
+                this.containerService.containerDetailResource.reload();
+              }
+            });
+          }
+        }
+      });
+  }
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/token/container-details/container-details-token-table/container-details-token-table.component.ts]: toggleAll and removeAll is not required here anymore as the token actions moved to a separate component

File: privacyidea/static_new/src/app/components/token/navigation/navigation.component.html [hunk 21]
```diff
@@ -342,15 +351,26 @@ <h4>{{ profileText }}</h4>
       </mat-list>
     </mat-expansion-panel>
 
-    <mat-expansion-panel disabled>
+    <mat-expansion-panel
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/components/token/navigation/navigation.component.html]: Please add the condition if the user has the right to see the policies

File: privacyidea/static_new/src/app/utils/parse-boolean-value.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+      false,
+      `Initial value for BoolSelectButtonsComponent must be "true" or "false" if string, but was ${initialValue}`
+    );
+  }
+  assert(
+    false,
+    `Initial value for BoolSelectButtonsComponent must be boolean, 0, 1, "true" or "false", but was ${initialValue}`
+  );
+  return false;
+}
```
Comment (jelinaunger) [inline privacyidea/static_new/src/app/utils/parse-boolean-value.ts:45]: What about tests for this function?

File: privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts [hunk 23]
```diff
@@ -0,0 +1,125 @@
+import { computed, signal } from "@angular/core";
+import { AuthData, AuthDetail, AuthRole, AuthServiceInterface, JwtData } from "../../app/services/auth/auth.service";
+import { MockAuthData, MockAuthDetail, MockPiResponse } from "../mock-services";
+import { of } from "rxjs";
+import { HttpHeaders } from "@angular/common/http";
+
+export class MockAuthService implements AuthServiceInterface {
```
Comment (jelinaunger) [inline privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts:26]: Still I would prefer implementing the mock services as child classes of the original service. This allows to use some original functions and properties. We than would only need to mock the http resources. 

For example, testing that some components / buttons are visible depending on the rights does not make any sense when mocking the full service. 

I know your argument of testing the component isolated since the service itself is already covered with its own tests is also valid. However, I think we than miss integration tests. Changes applied to the service might be forgotten to be applied to the mock class. Hence, all test might pass, cause the component works with the mock service but maybe not with the real service.

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html (hunk 8 above)
Comment (frankmer) [reply privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html:34]: Where do i overwrite the font here?

File: privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.ts (hunk 18 above)
Comment (frankmer) [reply privacyidea/static_new/src/app/components/policies/policy-panels/policy-panel/policy-panel.component.ts:119]: Added to the issue

File: privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts (hunk 23 above)
Comment (frankmer) [reply privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts:26]: I understand your point, but that's not how unit testing works.
In Unit Testing, it is best practice to mock all external dependencies (services, APIs, etc.) of the class you are testing. This isolates your code, making tests faster, repeatable, and ensures failures point directly to bugs in your class.
Maybe we could also write some integration tests where only the bare minimum is mocked.

File: privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html (hunk 8 above)
Comment (jelinaunger) [reply privacyidea/static_new/src/app/components/policies/policy-panels/action-tab/policy-description/policy-description.component.html:34]: Not sure, but obviously it is overwritten somewhere. However, this might not be crucial for now, we can also change this later.

File: privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts (hunk 23 above)
Comment (jelinaunger) [reply privacyidea/static_new/src/testing/mock-services/mock-auth-service.ts:26]: For correct unit testing you would also need to mock all other components.
Anyway, I think both approaches are valid, we just need to decide in general how we would like to test.

File: privacyidea/static_new/src/app/components/token/container-details/container-details-actions/container-details-actions.component.spec.ts (hunk 19 above)
Comment (frankmer) [reply privacyidea/static_new/src/app/components/token/container-details/container-details-actions/container-details-actions.component.spec.ts]: As actionAllowed is a mock, it has no real implementation and no longer depends on jwtData.
```
    authServiceMock.actionAllowed.mockImplementation((asd) => {
      console.warn(asd);
      return false;
    });
```
This was just a leftover from the debugging process. I changed it to:
```
    authServiceMock.actionAllowed.mockReturnValue(false);
```

**Comment by nilsbehlen** [comment]
there is overflow on the action name:
image

**Review by jelinaunger — APPROVED** [review]
